add doc for new delete option ignoreStoreReadErrorWithClusterBreakingPotential (#48463)
kubernetes/website: content/en/docs/reference/using-api/api-concepts.md

Once the last finalizer is removed, the resource is actually removed from etcd.

### Force deletion

{{< feature-state feature_gate_name="AllowUnsafeMalformedObjectDeletion" >}}

{{< caution >}}
This may break the workload associated with the resource being force deleted, if it
relies on the normal deletion flow, so cluster breaking consequences may apply.
{{< /caution >}}

By enabling the delete option `ignoreStoreReadErrorWithClusterBreakingPotential`, the

> user can perform an unsafe, and force **delete** operation of an undecryptable/corrupt

Reviewer: Is there missing a noun after `unsafe`?

We should explain why the operation is considered unsafe.

Reviewer 2: Maybe drop the _and_. I agree about explaining the safety risks more. I presume that this is because it can break invariants the control plane otherwise aims to maintain - have I got that right?

Reviewer 2: Also see my suggestion: https://github.com/kubernetes/website/pull/48463#discussion_r1854276913

Author: @stlaz I added `is unsafe because it ignores finalizer constraints, and skips precondition checks.`  and we also have the caution at the top, do you want the doc to mention any other specific concern?

> resource. This option is behind a new ALPHA feature gate, and it is disabled by

Reviewer 2: (see https://developers.google.com/style/timeless-documentation for why)
```suggestion
resource. This option is behind an alpha feature gate, and it is disabled by
```

default. In order to use this option, the cluster operator must enable the feature by
setting the command line option `--feature-gates=AllowUnsafeMalformedObjectDeletion=true`.

{{< note >}}
The user performing the force **delete** operation must have the
`unsafe-delete-ignore-read-errors` permission on the resource.
{{< /note >}}

A resource is considered corrupt if it can not be successfully retrieved from the

> storage due to a) transformation error e.g. decryption failure, or b) the object

Reviewer 2: We [avoid Latin phrases and their abbreviations](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-latin-phrases)
```suggestion
storage due to a) transformation error (for example: decryption failure), or b) the object
```

failed to decode. Normal deletion flow is attempted first, and if it fails with

> a `corrupt resource` error then it triggers the unsafe, and force delete. A force

Reviewer 2: ```suggestion
a _corrupt resource_ error then it triggers the unsafe mechanism, and forcibly deletes the data
from Kubernetes' backing store (etcd). A force
```
(unless the error code is literally the string `corrupt resource`, in which case the original is fine).

Author: the error message contains `corrupt object`. I changed it to _corrupt resource_

**delete** operation ignores finalizer constraints, and skips precondition checks.

> 

Reviewer 2: ```suggestion
Am I right that you cannot use `ignoreStoreReadErrorWithClusterBreakingPotential` to skip finalizers if the backing data are readable and could be removed without issue?

I think that's the case from what's written here, but I wanted to check.
```

Author: yes, you cannot avoid finalizers by setting the `ignoreStoreReadErrorWithClusterBreakingPotential` option, the apiserver will throw an error, I added a note.

The default value for this option is `false`, this maintains backward compatibility.

> When this field is set to `true`, the fields `dryRun`, `gracePeriodSeconds`,

Reviewer 2: ```suggestion
For **delete** requests with delete option field `ignoreStoreReadErrorWithClusterBreakingPotential` set to true,
the fields `dryRun`, `gracePeriodSeconds`,` `true`, the fields `dryRun`, `gracePeriodSeconds`,
```

Once the last finalizer is removed, the resource is actually removed from etcd.

> ### Force deletion

Reviewer 2: (nit)
```suggestion
### Forcible resource deletion {#forced-deletion}
```

Reviewer 3: should be `{#force-deletion}`, no `d`, to preserve existing links.

Reviewer 2: @tengqm this is new content (no existing links). Here I'm just recommending a way to make the eventual URL shorter.

{{< feature-state feature_gate_name="AllowUnsafeMalformedObjectDeletion" >}}

{{< caution >}}
This may break the workload associated with the resource being force deleted, if it
relies on the normal deletion flow, so cluster breaking consequences may apply.
{{< /caution >}}

By enabling the delete option `ignoreStoreReadErrorWithClusterBreakingPotential`, the
user can perform an unsafe, and force **delete** operation of an undecryptable/corrupt
resource. This option is behind an ALPHA feature gate, and it is disabled by
default. In order to use this option, the cluster operator must enable the feature by
setting the command line option `--feature-gates=AllowUnsafeMalformedObjectDeletion=true`.

{{< note >}}

> The user performing the force **delete** operation must have the
> `unsafe-delete-ignore-read-errors` permission on the resource.

Reviewer 2: (style nit)
```suggestion
The user performing the force **delete** operation must be permitted to perform
both the **delete** and **unsafe-delete-ignore-read-errors** verbs on that resource.
```

We write API verbs in non-monospaced bold. I assume that **unsafe-delete-ignore-read-errors** is a verb and that a SubjectAccessReview happens. Have I got that right?

Author: Yes, unsafe-delete-ignore-read-errors is a right/verb. I was going through the existing doc (for example https://kubernetes.io/docs/concepts/security/rbac-good-practices/) and how they format RBAC rights. I used the same format ( `verb` ) in this doc. It seems the format **delete** is used to represent an API operation, thoughts?

Reviewer 2: If you're defining access through RBAC, you use monospace. But: outside of the RBAC docs, we avoid assuming that clusters use RBAC for authz. It's a _common_ choice but not the only option.

Reviewer 2: I'm rewording here to avoid an RBAC-centric framing.

> In addition to `delete` rights, the user performing the force **delete**
> operation must have `unsafe-delete-ignore-read-errors` rights on the given resource.

Reviewer 2: (style nit)
```suggestion
The user performing the force **delete** operation must be permitted to perform
both the **delete** and **unsafe-delete-ignore-read-errors** verbs on that resource.
```

We write API verbs in non-monospaced bold. I assume that **unsafe-delete-ignore-read-errors** is a verb and that a SubjectAccessReview happens. Have I got that right?

Also, avoid implying that clusters always use Kubernetes RBAC. Many do but it's not mandatory to.
